### ISSUE TITLE
Problem with using `predict` with vector valued random variables

### ISSUE THREAD

**Issue by SamuelBrand1** [TuringLang/Turing.jl#2239]
Hi everyone,

## Problem

There seems to be a problem with using `predict` in conjunction with models that use vectorisation.

Consider this fairly simple example:

```math
\begin{split}
\sigma \sim \text{HalfNormal}(0.1) \\
\mu_i \sim \mathcal{N}(0, 1),\qquad i = 1,...,n \\
\epsilon_i \sim \mathcal{N}(0, \sigma^2),\qquad i = 1,...,n\\
x_i \sim \mu_i + \epsilon_i ,\qquad i = 1,...,n
\end{split}
```

We can generate a dataset by sampling from this model for (say) $n = 10$. The forecasting problem is then sampling for $n_f
 = 11,...,20$ (only information propagated forward is about variance of noise).

However, this fails as per below:

```julia
using Turing, StatsPlots, DynamicPPL, Random

Random.seed!(1234)

@model function mv_normal(n)
	σ ~ truncated(Normal(0., 0.1), lower = 0.)
	μ ~ MvNormal(n, 1.0) # Means
	x ~ MvNormal(μ, σ) # noise
	return x
end

mdl_10 = mv_normal(10)

# Sample data
x_data = mdl_10()

# infer means and obs noise
chn = sample(mdl_10 | (x = x_data,), NUTS(), 2_000)

# forecast
forecast_mdl = mv_normal(20)

forecast_chn = predict(forecast_mdl, chn; include_all = true)

let
	obs = generated_quantities(forecast_mdl, forecast_chn) |> X -> reduce(hcat, X)
	plt = plot(obs, c = :grey, alpha = 0.05, lab = "")
	scatter!(plt, x_data, c = :red, lab = "observed", title = "BAD FORECAST", ms = 6)
end
```
The failure mode here seems to be that the sample underlying random variables for $\epsilon_i,~ i = 11,...,20$ gets drawn across samples from from `chn`.

## Fix 1: `mapreduce` across `forecast` calls

So if you instead loop over samples and run `forecast` for each sample, this seems to work:

```julia
forecast_chn_mapreduce = mapreduce(vcat, 1:size(chn, 1)) do i
	c = predict(forecast_mdl, chn[i,:,1]; include_all = true)
        # Take care to set the range sequentially
	setrange(c, i:i)
end

let
	obs = generated_quantities(forecast_mdl, forecast_chn_mapreduce) |> X -> reduce(hcat, X)
	plt = plot(obs, c = :grey, alpha = 0.05, lab = "")
	scatter!(plt, x_data, c = :black, lab = "observed", title = "OK FORECAST")
end
```

## Fix 2: Non-vectorised sampling

Or you can modify the underlying model to not use calls to vectorised random variables (although IMO this is non-ideal).

```julia
@model function mv_normal_2(n)
	σ ~ truncated(Normal(0., 0.1), lower = 0.)
	μ = Vector{eltype(σ)}(undef, n)
	for i = 1:n
		μ[i] ~ Normal()
	end
	x ~ MvNormal(μ, σ) # noise
	return x
end

mdl2_10 = mv_normal_2(10)
x_data2 = mdl2_10()
chn2 = sample(mdl2_10 | (x = x_data2,), NUTS(), 2_000)


forecast_mdl2 = mv_normal_2(20)
forecast_chn2 = predict(forecast_mdl2, chn2; include_all = true)

let
	obs = generated_quantities(forecast_mdl2, forecast_chn2) |> X -> reduce(hcat, X)
	plt = plot(obs, c = :grey, alpha = 0.05, lab = "")
	scatter!(plt, x_data2, c = :black, lab = "observed", title = "ALSO OK FORECAST?")
end
```

## Ideal situation

Obviously, it would be ideal if `predict` "just worked" with vectorised random variables. Given the failure mode of naive usage of `predict` I'm assuming that this is a problem with how the random numbers are generated around [here](https://github.com/TuringLang/Turing.jl/blob/cb58871e321ef058d86b51adfe602a562bd690f4/src/inference/Inference.jl#L664)?

**Comment by SamuelBrand1** [comment]
Thanks for the detailed explanation!

So it turns out that this only works for models which have a representation where the priors are the same as the posterior.

TBF, this is actually a pretty large class of forecast models: discrete-time numerical solutions to SDEs and the finite dimensional distributions of a GP can be written this way (we were motivated by having a standard parameterisation of latent white noise).

**Comment by SamuelBrand1** [comment]
So to be clear, in the example where you fix _part of_ an array [here](https://turinglang.org/DynamicPPL.jl/stable/api/#DynamicPPL.fix) using the `Dict` form of fixing... that wouldn't work if the array had been declared by calling `~ MvNormal(...)`?

**Comment by torfjelde** [comment]
> TBF, this is actually a pretty large class of forecast models: discrete-time numerical solutions to SDEs and the finite dimensional distributions of a GP can be written this way (we were motivated by having a standard parameterisation of latent white noise).

Yeah, I definitively see the use for this! But it's somewhat non-trivial to support, so it really comes down to whether we want the maintenance burden of the functionality vs. having the user do some manual labour, i.e. use a `for` loop. 

> So to be clear, in the example where you fix part of an array [here](https://turinglang.org/DynamicPPL.jl/stable/api/#DynamicPPL.fix) using the Dict form of fixing... that wouldn't work if the array had been declared by calling ~ MvNormal(...)?

Exactly. 

The most annoying aprt of all this (IMO), is the perf implications of using vectorized vs. for loop. It's technically possible to do something like
```julia
if @performing_inference
    x ~ MvNormal(...)
else
    x = Vector(undef, 10)
    for i in eachindex(x)
        x[i] ~ Normal(...)
    end
end
```
but we don't have that implemented (related: https://github.com/TuringLang/DynamicPPL.jl/issues/510).

**Comment by torfjelde** [comment]
One simple approach that could also work is for the aformentioned code to be generated automatically through an `iid` macro or something, e.g.
```julia
@iid x ~ MvNormal(...)
```
and then this just converts to the above code block under the hood.

**Comment by SamuelBrand1** [comment]
Right. And this would avoid issues with (say) `adtype = AutoReverseDiff(true)` because grad calls only occur in "inference mode" so the existence of the slower performance branch wouldn't be relevant to sampling?

**Comment by torfjelde** [comment]
> Right. And this would avoid issues with (say) adtype = AutoReverseDiff(true) because grad calls only occur in "inference mode" so the existence of the slower performance branch wouldn't be relevant to sampling?

Exactly:)

**Comment by SamuelBrand1** [comment]
> > TBF, this is actually a pretty large class of forecast models: discrete-time numerical solutions to SDEs and the finite dimensional distributions of a GP can be written this way (we were motivated by having a standard parameterisation of latent white noise).
> 
> Yeah, I definitively see the use for this! But it's somewhat non-trivial to support, so it really comes down to whether we want the maintenance burden of the functionality vs. having the user do some manual labour, i.e. use a `for` loop.
> 
> > So to be clear, in the example where you fix part of an array [here](https://turinglang.org/DynamicPPL.jl/stable/api/#DynamicPPL.fix) using the Dict form of fixing... that wouldn't work if the array had been declared by calling ~ MvNormal(...)?
> 
> Exactly.
> 
> The most annoying aprt of all this (IMO), is the perf implications of using vectorized vs. for loop. It's technically possible to do something like
> 
> ```julia
> if @performing_inference
>     x ~ MvNormal(...)
> else
>     x = Vector(undef, 10)
>     for i in eachindex(x)
>         x[i] ~ Normal(...)
>     end
> end
> ```
> 
> but we don't have that implemented (related: [TuringLang/DynamicPPL.jl#510](https://github.com/TuringLang/DynamicPPL.jl/issues/510)).

Did anything happen on this? @seabbs and myself are trying to make something where you can fully compose a fairly large set of `Model`s defining different probabilistic components one might want in a fully feature epi model with `@submodel`. 

Since, prediction is quite important here it would be a handy feature to have, but otoh I'm not inclined to go through every single model definition and put in a "forecast mode" boolean switch. Given the known behaviour of `MvNormal` under conditioning some indices, couldn't you support that as a special case?

**Comment by seabbs** [comment]
Noting this also appears to be an issue with filldist(Normal(), n) which is Doc'd as idd so I assume it would work initially.

**Comment by seabbs** [comment]
I've started adding a PredictContext (+ hacking on `Turing.predict`) and have this switch-based approach on the way to working I think (I did the inverse of the suggestion as thought it would be easier for someone with no real understanding of the contexts system). Something that is very clear is that it is pretty clunky (even with passing around the if else block in a submodel) and the `@idd` macro suggestion would be much preferred.

**Comment by torfjelde** [comment]
Tagging @yebai to get some thoughts. Specifically on something like:

> One simple approach that could also work is for the aformentioned code to be generated automatically through an `iid` macro or something, e.g.
> 
> ```julia
> @iid x ~ MvNormal(...)
> ```
> 
> and then this just converts to the above code block under the hood.

**Comment by seabbs** [comment]
> If we fix filldist (and maybe the broadcasting syntax, e.g. μ .~ Normal()) would that be sufficient for this use case?

> Of course, we should provide a warning message to the user in all other unsupported cases discussed here.

Yes both of these would be great and cover our use case. It would also be nice to make it easier to switch modes if users find other edge cases they want to fix but we wouldn't need that for what we are doing (at least at the moment)

**Comment by SamuelBrand1** [comment]
> @seabbs @SamuelBrand1, we would be happy to hear more of your thoughts on syntax design. I think it has to be robust and intuitive for long-term maintenance.
> 
> EDIT:
> 
> > Noting this also appears to be an issue with filldist(Normal(), n) which is Doc'd (i think) as idd so I assumed it would work initially.
> 
> If we fix `filldist` (and maybe the broadcasting syntax, e.g. `μ .~ Normal()`) would that be sufficient for this use case?
> 
> Of course, we should provide a warning message to the user in all other _unsupported_ cases discussed here.

This sounds good, but would there be a performance implication? I'm wondering about the upsides/downsides here.

**Comment by torfjelde** [comment]
> Noting this also appears to be an issue with filldist(Normal(), n) which is Doc'd (i think) as idd so I assumed it would work initially.

IMO this is more of a doc-issue, as there are definitively scenarios where you it makes sense to use `filldist` rather than `arraydist`, as the former is more efficient (when applicable).

> μ .~ Normal()) would that be sufficient for this use case?

It's somewhat unclear to me how this would work, but maybe this is something we should discuss in DynamicPPL.jl :)

**Comment by seabbs** [comment]
> IMO this is more of a doc-issue, as there are definitively scenarios where you it makes sense to use filldist rather than arraydist, as the former is more efficient (when applicable).

My point was that `filldist` also doesn't work with predict at the moment. 

> It's somewhat unclear to me how this would work, but maybe this is something we should discuss in DynamicPPL.jl :)

What are the steps forward from here? In principle happy to help out with a proposed fixed but likely to need guidance / its unclear how y'all manage community contributions. 

It would be really great for us to have this fixed centrally vs implementing modifications in our tooling to get this working.

**Comment by seabbs** [comment]
Ah I see discussion over on https://github.com/TuringLang/DynamicPPL.jl/issues/510 so I guess I'll track that!

**Comment by yebai** [comment]
This is best treated as a **documentation issue** rather than a bug: `predict` is not intended to be used this way, and the fixed-array case working at all is incidental.

The reason is a matter of semantics. Turing treats each variable exactly as it is written in the model. A statement like `μ ~ MvNormal(zeros(n), 1)` declares a **single** `n`-dimensional random variable — not `n` i.i.d. scalars. Consequently, `predict` can condition on `μ` or resample it, but it cannot condition on *part* of it and resample the remainder, because that operation is a marginalization, which Turing does not perform.

**Why the forecast is a misuse.** The chain is fit with `n = 10` (so it carries `μ[1:10]`), and `predict` is then called on the `n = 20` model, where `μ` is one 20-dimensional variable:

```julia
@model function mv_normal(n)
    σ ~ truncated(Normal(0, 0.1); lower = 0)
    μ ~ MvNormal(zeros(n), 1)   # ONE 20-dim variable when n = 20
    x ~ MvNormal(μ, σ)
    return x
end

chn = sample(mv_normal(10) | (x = x_data,), NUTS(), 2_000)
predict(mv_normal(20), chn)     # asks predict to fix μ[1:10] and resample μ[11:20]
```

There is no supported way to fix ten components of a twenty-dimensional variable, so the result is not meaningful (in older versions `μ[11:20]` was silently frozen to a single prior draw; the behaviour is version-dependent and should not be relied on).

**Recommended workaround.** Declare the variable as i.i.d. components with an explicit index, so that each `μ[i]` is its own random variable that `predict` can treat independently:

```julia
@model function mv_normal(n)
    σ ~ truncated(Normal(0, 0.1); lower = 0)
    μ = Vector{typeof(σ)}(undef, n)
    for i in 1:n
        μ[i] ~ Normal()
    end
    x ~ MvNormal(μ, σ)
    return x
end
```

With this formulation `μ[1:10]` are taken from the chain and `μ[11:20]` are drawn from the prior for each posterior sample, as intended for a forecast.

**Comment by seabbs** [comment]
> o that each μ[i] is its own random

That is a shame this approach is very slow in inference and so means you need different inference and prediction versions of the models.

**Comment by SamuelBrand1** [comment]
> **Recommended workaround.** Declare the variable as i.i.d. components with an explicit index, so that each `μ[i]` is its own random variable that `predict` can treat independently:
> 
> @model function mv_normal(n)
>     σ ~ truncated(Normal(0, 0.1); lower = 0)
>     μ = Vector{typeof(σ)}(undef, n)
>     for i in 1:n
>         μ[i] ~ Normal()
>     end
>     x ~ MvNormal(μ, σ)
>     return x
> end
> With this formulation `μ[1:10]` are taken from the chain and `μ[11:20]` are drawn from the prior for each posterior sample, as intended for a forecast.

That is a pretty painful workaround for the reason that @seabbs just put up. 

Is there any any associated work on smoothing out any of:

- The performance regression of doing index by index independents rather than multivariate/`filldist`.
- Allowing marginalisation on specific multivariate dists where its trival e.g. MvNormal which I guess is the most popular mv dist?

**Comment by yebai** [comment]
> Allowing marginalisation on specific multivariate dists where its trival e.g. MvNormal which I guess is the most popular mv dist?

For univariate IID random variables and MvNormal, closed-form marginalisation is possible. That would be a messy partial solution. 

I'd encourage you to perform some benchmarking to see how much of a performance difference there is between the loop and the vectorised version. 

EDIT: please open a new issue towards DynamicPPL if needed.
